src/buffer_par.rs

```diff
@@ -0,0 +1,523 @@
+use rayon::iter::plumbing::*;
+use rayon::iter::{IndexedParallelIterator, ParallelIterator};
+use rayon::slice::{ChunksExact, ChunksExactMut, ParallelSlice, ParallelSliceMut};
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+
+use crate::traits::Pixel;
+use crate::ImageBuffer;
+
+/// Parallel iterator over pixel refs.
+pub struct PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    chunks: ChunksExact<'a, P::Subpixel>,
+}
+
+impl<'a, P> ParallelIterator for PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    type Item = &'a P;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .drive_unindexed(consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.len())
+    }
+}
+
+impl<'a, P> IndexedParallelIterator for PixelsPar<'a, P>
+where
+    P: Pixel + Sync + 'a,
+    P::Subpixel: Sync + 'a,
+{
+    fn drive<C: Consumer<Self::Item>>(self, consumer: C) -> C::Result {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .drive(consumer)
+    }
+
+    fn len(&self) -> usize {
+        self.chunks.len()
+    }
+
+    fn with_producer<CB: ProducerCallback<Self::Item>>(self, callback: CB) -> CB::Output {
+        self.chunks
+            .map(|v| <P as Pixel>::from_slice(v))
+            .with_producer(callback)
+    }
+}
+
+impl<P> Clone for PixelsPar<'_, P>
```

Comment:
Can this implementation be derived automatically?

Reply:
It looks like `derive(Clone)` on this automatically puts a `Clone` bound on `P`. Since `Pixels`/`EnumeratePixels` doesn't have this bound, I assume it's unnecessary.

Reply:
The `P: Pixel` bounds implies `P: Clone` (see https://docs.rs/image/latest/image/trait.Pixel.html)